entry/src/block_component.rs

```diff
@@ -0,0 +1,2165 @@
+    /// # Errors
+    /// Returns a bincode error if serialization fails or if validation fails.
+    pub fn to_bytes(&self) -> Result<Vec<u8>, bincode::Error> {
+        let mut buffer = Vec::new();
+
+        match self {
+            Self::Entries(entries) => {
+                Self::validate_entries_length(entries.len())
+                    .map_err(|e| bincode::Error::new(bincode::ErrorKind::Custom(e)))?;
+
+                // Write entry count
+                buffer.extend_from_slice(&(entries.len() as u64).to_le_bytes());
+
+                // Write each entry
+                for entry in entries {
+                    let entry_bytes = bincode::serialize(entry)?;
+                    buffer.extend_from_slice(&entry_bytes);
+                }
+            }
+            Self::BlockMarker(marker) => {
+                // Write zero entry count
+                buffer.extend_from_slice(&0u64.to_le_bytes());
+                // Write marker data
+                let marker_bytes = marker.to_bytes()?;
+                buffer.extend_from_slice(&marker_bytes);
+            }
+        }
+
+        Ok(buffer)
+    }
+
+    /// Deserializes from bytes with validation.
+    ///
+    /// # Errors
+    /// Returns a bincode error if deserialization fails or data is invalid.
+    ///
+    /// TODO(karthik): fuzz test this function.
+    /// TODO(karthik): should we error out if there are remaining bytes after deserialization?
+    pub fn from_bytes(data: &[u8]) -> Result<Self, bincode::Error> {
+        const ENTRY_COUNT_SIZE: usize = 8;
+
+        let entry_count = u64::from_le_bytes(
+            data.get(..ENTRY_COUNT_SIZE)
+                .and_then(|bytes| bytes.try_into().ok())
+                .ok_or_else(|| bincode::Error::new(bincode::ErrorKind::SizeLimit))?,
+        );
+
+        // Validate entry count
+        if entry_count as usize >= Self::MAX_ENTRIES {
+            return Err(bincode::Error::new(bincode::ErrorKind::Custom(format!(
+                "BlockComponent entry count {entry_count} exceeds maximum {}",
+                Self::MAX_ENTRIES
+            ))));
+        }
+
+        let mut cursor = ENTRY_COUNT_SIZE;
+        let mut entries = Vec::with_capacity(entry_count as usize);
+
+        // Deserialize entries
+        for _ in 0..entry_count {
+            let remaining_data = data
+                .get(cursor..)
+                .ok_or_else(|| bincode::Error::new(bincode::ErrorKind::SizeLimit))?;
+
+            let entry: Entry = bincode::deserialize(remaining_data)?;
+            let entry_size = bincode::serialized_size(&entry)? as usize;
+            entries.push(entry);
+            cursor += entry_size;
+        }
+
+        // Handle remaining data
+        match data.get(cursor..) {
+            Some(remaining) if !remaining.is_empty() => {
+                // Remaining data should be a marker
+                let marker = VersionedBlockMarker::from_bytes(remaining)?;
+                if entries.is_empty() {
```

Comment:
think it would be better to case on whether entry_count == 0 and deserialize accordingly. Cant' have both in the same entry batch.